services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java

```diff
@@ -956,7 +960,7 @@ public static Response redirectAfterSuccessfulFlow(KeycloakSession session, Real
             clientSession.setNote(SSO_AUTH, "true");
             authSession.removeAuthNote(SSO_AUTH);
         } else {
-            int authTime = Time.currentTime();
+            int authTime = Optional.ofNullable(authSession.getClientNote(AUTH_TIME_BROKER)).map(Integer::parseInt).orElse(Time.currentTime());
             userSession.setNote(AUTH_TIME, String.valueOf(authTime));
```

Comment:
Here when the `Pass Max Age` is enabled but `max_age` is not passed( this is especially the case when the user logout from Keycloak and log in again) the value for  `auth_time` will be updated, I think it should not happen, let's see how clients do the checks for sensitive action:

```
public boolean isSensitiveActionAllowed() {
  long authTime = Optional.ofNullable(oidcHelper.getAuthTime()).orElse(0);
  long now = Time.currentTime();
  return now - authTime < sensitiveActionConfig.getSensitiveActionAllowedDuration();
}

public void sensitiveAction() {
  if (isSensitiveActionAllowed())
  // Do the sensitive action
  else
  // redirect the user for reauthentication and then retry this action.
}

```

The literal description of this sample code:

> When the user tries to do the sensitive action and the `auth_time` is not present in the token, or it is old, the client redirects the user for re-authentication, otherwise, they let the user do the sensitive action.

Now imagine the `auth_time` is old and the user should not be able to do the sensitive action. But Instead of re-authentication, a hacker just logs out from the Keycloak, and logs in again(without going through the natural flow of re-authentication), when they log in again they will have updated `auth_time` and they can do the sensitive action.

I understand that this is complex but if the user can do the sensitive action without reauthentication all this effort would be useless.

Solutions:

* Not to pass `auth_time` to the client when the `Pass Max Age` is enabled and Keycloak does not know the authentication time in IDP(since the `max_age` param is not passed to Keycloak) it might not be backward compatible because Keycloak always passed `auth_time` in the token up to now(But protocol said that it is only required when `max_age` is passed)

*  Return `0` for `auth_time` when Keycloak does not know authentication time in IDP(When `Pass Max Age` is ON and `max_age` is not passed)

Reply:
> Now imagine the auth_time is old and the user should not be able to do the sensitive action. But Instead of re-authentication, a hacker just logs out from the Keycloak, and logs in again(without going through the natural flow of re-authentication), when they log in again they will have updated auth_time and they can do the sensitive action.

You are assuming here that the account is compromised, right? If so, there are many other things like completely stealing the account.

I think what you want to do here is make sure re-authentication happens after logout. In this case, we could perhaps consider a specific claim within the token to indicate that authentication happened at the IdP.

W.r.t. to the proposed solutions:

1) The `auth_time` is an optional claim and I think it makes sense to include it even though you are not using `max_age`.
2) Looks like we are overusing the claim
3) For that we would need to persist the `auth_time` so they can survive session expiration and server restarts. Not sure if we want that.

If it makes sense for you, I would keep the scope of this PR to only include the validation of `auth_time` and the latest change to propagate the time from the OP. We can then create a follow-up issue and discuss the requirements you are proposing.

Reply:
First of all thanks for the effort in this PR, I thought about it for a long time and I know how complex it is.

> You are assuming here that the account is compromised, right? If so, there are many other things like completely stealing the account.

No, in this scenario, the account is not compromised. Just another malicious user uses our user device and does the sensitive action.  The goal of re-authentication is that the user who is doing the action has the `username` and `password` and thereby he is the identified user. if someone else is using the same device and do not have the username and password should not be able to do the sensitive action.

Please Consider UserA as the owner of the account, and UserB as a malicious user who finds out that UserA is not at his desk and wants to use UserA device and UserA account to do some malicious action.

Here is the scenario for reauthentication that is implemented in this PR.

* UserA wants to do sensitive actions.
*  Client checks the `auth_time` and based on that redirects the user to Keycloak for reauthentication.
*  User clicks on the IDP button.
* Keycloak redirects the user to IDP for reauthentication.
*  User provides its credential(Maybe username and password). 
* User redirects back to Keycloak -> user redirects back to the client.
*  client checks the `auth_time` and lets the sensitive action.

Here is the scenario in which UserB gets at the UserA desk and wants to do some sensitive action, so considering this all the following actions are done with userA device and using userA session with Keycloak and upstream IDP.

* UserB wants to do sensitive action.
* Client checks the `auth_time` and redirects the user for re-authentication.
* At this point, UserB logout from Keycloak(He only logs out from Keycloak, and he is still logged in upstream IDP like Azure AD).
*  Then he login again to Keycloak.
*  He clicks on IDP button. 
*  He will be redirected to IDP.
* **IDP redirects userB back to Keycloak(Note that this time UserB does not need to provide userA credentials,  and  IDP does not ask for the username and password because the UserA session with upstream IDP is still open and the `max_age` param is not provided).**
*  Keycloak redirects back the user to the client with updated `auth_time`.
*  UserB goes back to the sensitive action and does that action, this time he can do that action because he has updated `auth_time` from the client's perspective.

Reply:
Thanks. It is clear now the attack scenario you are considering.

I'm not sure how others OPs deal with both `acr` and `amr` claims. This is also somewhat related to step-up. My point is, a `acr==0` might indicate that the authentication was done based on long-lived cookies and as such your client could deny the sensitive operation.

I don't max-age alone will solve this but perhaps a mix of checking `acr`, conditionally forcing `prompt=login`, etc.

There is also the possibility to store tokens when doing brokering. We can also think about comparing the stored token with the new token issued by the OP.

Reply:
Thank you so much for the answer, it would be also appreciated if you can create an issue to discuss it.

I also thought about step-up authentication but the question that might be raised there is that:

When we are using upstream IDP like azure AD, is that OK to use second-factor authentication in Keycloak, or when the user is being authenticated using upstream IDP all the things that are related to authentication should be done in upstream IDP?

Reply:
Hard to say because it depends on your requirements. If the OP is not enough for you in terms of proof of identity or if you want to also enable users to authenticate through your OP, the it makes sense.

But I think your use case is more related to step-up and how brokering integrates with this capability so that LoAs are properly set when authentication happens at the brokered OP.

Reply:
I agree that it is a complex problem and we can possibly address as a follow-up.

In general, it seems non-trivial to decide what should be proper `auth_time` sent to the client. It is quite clear that in case Keycloak shows interactive screens to the user during auth (EG. username/password, TOTP, WebAuthn), the `auth_time` should be updated. But it is not clear if/how it should be set in case of non-interactive authenticators like Kerberos or X509. And for brokering, where Keycloak delegates authentication to IDP, it is also not clear.

I wonder that we can possibly have `auth_time` protocol mapper? This will allow to configure if:
- The `auth_time` should be set just if interactive authentication was triggered (It is possible to detect interactive authenticators as the authenticators like Password, OTP or WebAuthn implements `CredentialValidator` interface. And the info about authenticators used in finished authentication flow is saved in the authenticationSession)
- The `auth_time` should be set by any authenticator (with the exception of SSO Cookie authenticator)
- The `auth_time` should be set based on some note from authentication session. So the note can be set by the IDP (like done in this PR) or by some authenticator (or by authentication policy in the future)
- Maybe more options (or combinations of the options above).

This may allow to set that for example if authentication happened through Kerberos OR if authentication happened through IDP and IDP did not returned `auth_time`, then the `auth_time` should not be sent at all in the ID Token etc.

Until we have, I am not sure if we should skip sending `auth_time` to client (or set `auth_time` to 0). So I will likely rather keep as it is in this PR.

So +1 to discuss as a follow-up.